src/Validation/Validator.php

```diff
@@ -635,8 +706,92 @@ public function requirePresence($field, $mode = true, $message = null)
      * @return $this
      */
     public function allowEmpty($field, $when = true, $message = null)
+    {
+        return $this->allowEmptyByFlags($field, null, $when, $message);
+    }
+
+    /**
+     * Allows a field to be empty. You can also pass array.
+     * Using an array will let you provide the following keys:
+     *
+     * - `flags` individual flags for field
+     * - `when` individual when condition for field
+     * - 'message' individual message for field
+     *
+     * You can also set flags, when and message for all passed fields, the individual
+     * setting takes precedence over group settings.
+     *
+     * ### Example:
+     *
+     * ```
+     * // Email can be empty
+     * $validator->allowEmptyByFlags('email', Validator::EMPTY_STRING);
+     *
+     * // Email can be empty on create
+     * $validator->allowEmptyByFlags('email', Validator::EMPTY_STRING, 'create');
+     *
+     * // Email can be empty on update
+     * $validator->allowEmptyByFlags('email', Validator::EMPTY_STRING, 'update');
+     *
+     * // Email and subject can be empty on update
+     * $validator->allowEmptyByFlags(['email', 'subject'], Validator::EMPTY_STRING, 'update');
+     *
+     * // Email can be always empty, subject and content can be empty on update.
+     * $validator->allowEmptyByFlags(
+     *      [
+     *          'email' => [
+     *              'when' => true
+     *          ],
+     *          'content' => [
+     *              'message' => 'Content cannot be empty'
+     *          ],
+     *          'subject'
+     *      ],
```

Comment:
Do we need to support this mode? It seems awkward to use.

Reply:
We could deprecate it in 3.7.

Reply:
Could we contain this mode to only exist in the older `allowEmpty()`? That would allow us to never add it avoiding a method that is deprecated/removed right away.

Reply:
We could. Should I fix doc comments of type-specific allowEmpty*()s? Or should I throw an exception in those methods?

Reply:
I would fix the doc comments for now. We can add stricter types in 4.0

Reply:
Ok.

Reply:
I am OK with reverting allowEmpty() to the original state. then we would not need to care about the backward compatibility.

Reply:
I agree with mark, this is better if this is just an internal method.

Reply:
I have already explained why I made it as public. But I don't stick with my thought. Feel free to add your commits.

Reply:
@chinpei215 I'll push some commits up tonight.

Reply:
I pushed up my change to remove the array field support from new methods.